content/posts/creating-grafana-dashboards-for-nodejs-on-kubernetes/index.md

```diff
@@ -0,0 +1,261 @@
+
+The Express Prometheus Bundle automatically exports a number of useful metrics related to request duration and is based on [prom-client](https://github.com/siimon/prom-client) which you can use to export additional custom metrics for your application.
+
+## Example Application
+
+We're working with bespinian's Express based [Awesome Node](https://github.com/bespinian/awesome-node) example application. It's deployed into a namespace called `awesome-node`, exports metrics using the Express Prometheus Bundle and is running 3 replicas.
+
+## Requests Per Second
+
+Let's start creating some diagrams. In a first panel, you may want to show a graph, plotting requests per second to your application over time such as the one below.
+
+![Graph showing requests per second over time](./requests-per-second.png)
+
+This Grafana panel shows requests per second hitting all instances of the application over 30 minutes using the following query.
+
+```promql
+sum by (app) (rate(http_request_duration_seconds_count{app="awesome-node"}[2m]))
+```
+
+The query is retrieving a range vector over 2 minutes with data points of the `http_request_duration_seconds_count` metric with the label `app` set to `awesome-node`.
+In our example Prometheus scrapes the metrics every minute, so with every two minutes interval there should be between 1 and 3 data points.
+
+The `rate()` function then calculates the average rate per second with respect to the exact intervals between data points.
+
+This still results in multiple metrics which are different in their `satus_code` and `kubernetes_pod_name` labels and can be summed up to give a single number for the whole applications.
+
+To display a graph over time, Grafana makes a query range over 30 minutes with 15 second intervals (these values can be configured differently in Grafana). Prometheus then calculates the query for each of those intervals, allowing Grafana to plot a chart.
+
+### The Counter Metric Type
+
+In this graph, we used the counter metric type, which is monotonically increasing. With such metrics we are usually interested in the the relative increase in a range rather than in their absolute value. The `rate()` function calculates this increase and takes into account the exact time between each two data points. It also accounts for counter resets, which may happen when an instance restarts.
+
+### The Gauge Metric Type
+
+Unlike counters, gauges are metrics that can also decrease in value. As such they are very intuitive to use and do not require derivation or the use of the `rate()` function.
+Examples for gauges could be temperature or CPU load. However, measuring CPU load using gauges may not be advised, since the gauge only represents a snapshot at sampling time and does not take into account any variation between two samples. Therefore tools like Prometheus Node Exporter export CPU usage in seconds as a counter type metric.
```

Comment:
New paragraphs within markdown that are not preceded with a blank line do not serve any purpose and are rather confusing because they might suggest that the finally compiled HTML will reflect that new paragraph. I would either omit them or use a blank line to create a new paragraph in the HTML.